odoo/addons/base/transpiler/transpiler_js.py

```diff
@@ -1,378 +1,393 @@
+the original source need to be supported by the browsers.
+"""
+
+
+def convert_odoo_module(url, content):
+    """
+    Transpile the code through a bunch of operations.
+    The order of the operations does sometimes matter.
+
+    :param content: The original source code
+    :param url: The url of the file in the project
+    :return: The transpiled source code
+    """
+    module_path = url_to_module_path(url)
+    legacy_odoo_define = get_aliased_odoo_define_content(module_path, content)
+
+    new_content = convert_legacy_default_import(content)
+    new_content = convert_basic_import(new_content)
+    new_content = convert_default_import(new_content)
+    new_content = convert_star_import(new_content)
+    new_content = convert_unnamed_relative_import(new_content)
+    new_content = convert_from_export(new_content)
+    new_content = convert_star_from_export(new_content)
+    new_content = convert_relative_require(url, new_content)
+    new_content = remove_index(new_content)
+    new_content = convert_function_and_class_export(new_content)
+    new_content = convert_variable_export(new_content)
+    new_content = convert_list_export(new_content)
+    new_content = convert_default(new_content)
+    new_content = wrap_with_odoo_define(module_path, new_content)
+    if legacy_odoo_define:
+        new_content += legacy_odoo_define
+    return new_content
+
+
+def url_to_module_path(url):
+    """
+    Odoo modules each have a name. (odoo.define("<the name>", function (require) {...});
+    It is used in to be required later. (const { something } = require("<the name>").
+    The transpiler transforms the url of the file in the project to this name.
+    It takes the module name and add a @ on the start of it, and map it to be the source of the static/src (or
+    static/tests) folder in that module.
+
+    in: web/static/src/one/two/three.js
+    out: @web/one/two/three.js
+    The module would therefore be defined and required by this path.
+
+    :param url: an url in the project
+    :return: a special path starting with @<module-name>.
+    """
+    match = re.match(r"""/?(?P<module>\w+)/([\w/]*/)?static/(?P<type>src|tests)(?P<url>/[\w/]*)""", url)
+    # result = re.match(r"""
+    #    /?                 # Optional /
+    #    (?P<module>        # Capture in a group named module
+    #        \w+            # One or more word character
+    #    )
+    #    /                  # /
+    #    ([\w/]*/)?         # A/few word character or /, zero to many times ended by /
+    #    static             # static
+    #    /                  # /
+    #    (?P<type>          # Capture in a group named type
+    #        src|tests      # src or tests
+    #    )
+    #    (?P<url>           # Capture in a group named url
+    #        /[\w/]*        # A word character or /, zero to many times.
+    #    )
+    #    """, url, re.VERBOSE)
```

Comment:
Still wondering why you're using the single-line version rather than the VERBOSE version.

Though I don't really see the point of the comments literally repeating what the regex says, I'd suggest explaining things at a higher level e.g. how the pattern makes sense or corresponds to the group.

Reply:
It was a test to go towards this kind of documentation: https://medium.com/@awlucattini_60867/writing-readable-regexes-fa7433829b7f
I like the idea of providing a link to regex101 with examples and counter examples. But I'm not a fan of the verbose mode. I feel that the regex are readable as they stand. But maybe juste a few comments with verbose mode like you propose could be the best option.

Reply:
> It was a test to go towards this kind of documentation: medium.com/@awlucattini_60867/writing-readable-regexes-fa7433829b7f

That seems completely useless to me. Unless you have a very complicated pattern this is as useful as defining `a_int = 1`. We know it's an int, it's written on the tin. Documenting the components at a high level e.g.

```python
pattern = re.compile(r"""
                      ^
                      \S+           # Prefix word
                      \ 
                      \$            # Currency sign
                      (\d+\.\d{2})  # monetary value
                      $
                      """, re.VERBOSE)
```
I find useful (at least in more complicated regex than this one), because it tells me *what* each segment is trying to match, which can be unclear / confusing especially with alternations or optional parts.

But you're using named groups which largely obviate comments (at least for the bits which do get collected but they tend to be the ones more in need of documentation). Though even then it can be useful to use verbose mode to make the regex less *dense*, a bit of spacing can make the boundaries of groups clearer and increase the readability in the same way paragraphs do in normal text.

> But I'm not a fan of the verbose mode. I feel that the regex are readable as they stand.

I'm ok with that, but having the verbose version commented out next to the non-verbose version is confusing. I think my original comment was to remove either.